File: javascript/src/stacks/0155-min-stack.js
const MinStack = function () {
  this.stack = [];
  this.min = [];
};

/**
 * @param {number} val
 * @return {void}
 *
 */
MinStack.prototype.push = function (val) {
  if (this.stack.length === 0 && this.min.length === 0) {
    this.stack.push(val);
    this.min.push(val);
  } else {
    const currentMin = this.min[this.min.length - 1];
    this.stack.push(val);
    this.min.push(Math.min(currentMin, val));
  }
};

/**
 * @return {void}
 *
 *
 */
MinStack.prototype.pop = function (val) {
  this.stack.pop();
  this.min.pop();
};

/**
 * @return {number}
 */
MinStack.prototype.top = function () {
  if (this.stack.length > 0) {
    return this.stack[this.stack.length - 1];
  }
};

/**
 * @return {number}
 */
MinStack.prototype.getMin = function () {
  if (this.min.length > 0) {
    return this.min[this.min.length - 1];
  }
};

module.exports = MinStack;

File: golang/stacks/0155-min-stack.go
package stacks

type MinStack struct {
	min   []int
	stack []int
}

func Constructor() MinStack {
	return MinStack{}
}

func (this *MinStack) Push(val int) {
	this.stack = append(this.stack, val)

	if len(this.min) > 0 {
		this.min = append(this.min, min(val, this.min[len(this.min)-1]))
	} else {
		this.min = append(this.min, val)
	}
}

func (this *MinStack) Pop() {
	this.stack = this.stack[:len(this.stack)-1]
	this.min = this.min[:len(this.min)-1]
}

func (this *MinStack) Top() int {
	return this.stack[len(this.stack)-1]
}

func (this *MinStack) GetMin() int {
	return this.min[len(this.min)-1]
}
